src/aten/sycl/DistributionTemplates.h

```diff
@@ -0,0 +1,595 @@
+struct UniformIntFullRangeTransformFunctor {
+  auto operator()(V val) const {
+    return at::transformation::uniform_int_full_range<T, V>(val);
+  }
+};
+
+template <typename RNG>
+void random_full_64_bits_range_kernel(TensorIteratorBase& iter, RNG gen) {
+  AT_DISPATCH_ALL_TYPES_AND(
+      at::ScalarType::BFloat16,
+      iter.dtype(),
+      "random_full_64_bits_range_kernel_xpu",
+      [&] {
+        if (std::is_same<scalar_t, int64_t>::value ||
+            std::is_same<scalar_t, double>::value ||
+            std::is_same<scalar_t, float>::value ||
+            std::is_same<scalar_t, at::BFloat16>::value) {
+          distribution_nullary_kernel<
+              scalar_t,
+              uint64_t,
+              rand4_engine_calls / 2>(
+              iter,
+              gen,
+              Rand4ULL2DistFunctor(),
+              UniformIntFullRangeTransformFunctor<scalar_t, uint64_t>());
+        } else {
+          TORCH_CHECK(
+              false,
+              "random_full_64_bits_range_kernel_xpu handles only int64, double, float and bfloat16");
+        }
+      });
+}
+
+template <typename T, typename V>
+struct UniformIntTransformFunctor {
+  auto operator()(V val) const {
+    return at::transformation::uniform_int<T, V>(val);
+  }
+};
+
+template <typename RNG>
+void random_kernel(TensorIteratorBase& iter, RNG gen) {
+  AT_DISPATCH_ALL_TYPES_AND3(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      at::ScalarType::Bool,
+      iter.dtype(),
+      "random_kernel_xpu",
+      [&] {
+        if (std::is_same<scalar_t, double>::value ||
+            std::is_same<scalar_t, int64_t>::value) {
+          distribution_nullary_kernel<
+              scalar_t,
+              uint64_t,
+              rand4_engine_calls / 2>(
+              iter,
+              gen,
+              Rand4ULL2DistFunctor(),
+              UniformIntTransformFunctor<scalar_t, uint64_t>());
+        } else {
+          distribution_nullary_kernel<scalar_t, uint32_t, rand4_engine_calls>(
+              iter,
+              gen,
+              Rand4DistFunctor(),
+              UniformIntTransformFunctor<scalar_t, uint32_t>());
+        }
+      });
+}
+
+// ============================================
+
+struct Uniform2DistributionFunctor {
+  auto operator()(randStatePhilox4_32_10_t* state) const {
+    return rand_uniform2_double(state);
+  }
+};
+
+struct Uniform4DistributionFunctor {
+  auto operator()(randStatePhilox4_32_10_t* state) const {
+    return rand_uniform4(state);
+  }
+};
+
+template <
+    typename scalar_t,
+    typename accscalar_t,
+    size_t engine_calls,
+    typename RNG,
+    typename transform_t>
+void uniform_and_transform(
+    TensorIteratorBase& iter,
+    RNG gen,
+    transform_t transform) {
+  // Distribution backbone only handle two accumulate type.
+  if (std::is_same<scalar_t, double>::value) {
+    Uniform2DistributionFunctor f;
+    distribution_nullary_kernel<scalar_t, accscalar_t, engine_calls / 2>(
```

Comment:
What is our `curand_uniform2_double`?
![image](https://github.com/intel/torch-xpu-ops/assets/28250760/9b606294-a396-45a1-9ddb-f9a2b3506fd5)